cabal-install/Distribution/Client/Store.hs

```diff
@@ -0,0 +1,207 @@
+
+
+-- | Return the 'UnitId's of all packages\/components already installed in the
+-- store.
+--
+getStoreEntries :: StoreDirLayout -> CompilerId -> Rebuild (Set UnitId)
+getStoreEntries StoreDirLayout{storeDirectory} compid = do
+    paths <- getDirectoryContentsMonitored (storeDirectory compid)
+    return $! mkEntries paths
+  where
+    mkEntries     = Set.delete (mkUnitId "package.db")
+                  . Set.delete (mkUnitId "incoming")
+                  . Set.fromList
+                  . map mkUnitId
+                  . filter valid
+    valid ('.':_) = False
+    valid _       = True
+
+
+newStoreEntry :: Verbosity -> StoreDirLayout
+              -> CompilerId -> UnitId
+              -> (FilePath -> IO a) -> IO () -> IO a
+newStoreEntry verbosity storeDirLayout@StoreDirLayout{..}
+              compid unitid
+              copyFiles register =
+    withTempIncomingDir storeDirLayout compid $ \incomingEntryDir -> do
+
+      -- Write all store entry files into the temp dir.
+      x <- copyFiles incomingEntryDir
+
+      -- Take a lock named after the 'UnitId' in question.
+      withIncomingUnitIdLock storeDirLayout compid unitid $ do
+
+        -- Check for the existence of the final store entry directory.
+        exists <- doesStoreEntryExist storeDirLayout compid unitid
+
+        if exists
+          -- If the entry exists then we lost the race and we must abandon,
+          -- unlock and re-use the existing store entry.
+          then do
+            removeDirectoryRecursive incomingEntryDir
+
+            debug verbosity $
+                "Concurrent build race: abandoning build in favour of existing "
+             ++ "store entry " ++ display compid </> display unitid
+
+          -- If the entry does not exist then we won the race and can proceed.
+          else do
+
+            -- Register the package into the package db (if appropriate).
+            register
+
+            -- Atomically rename the temp dir to the final store entry location.
+            renameDirectory incomingEntryDir finalEntryDir
+
+            debug verbosity $
+              "Installed store entry " ++ display compid </> display unitid
+
+      return x
+  where
+    finalEntryDir = storePackageDirectory compid unitid
+
+
+withTempIncomingDir :: StoreDirLayout -> CompilerId
+                    -> (FilePath -> IO a) -> IO a
+withTempIncomingDir StoreDirLayout{storeIncomingDirectory} compid action = do
+    createDirectoryIfMissing True (storeIncomingDirectory compid)
+    bracketOnError
+      (createTempDirectory (storeIncomingDirectory compid) "new")
+      (handleDoesNotExist () . removeDirectoryRecursive)
+      action
+
+
+withIncomingUnitIdLock :: StoreDirLayout -> CompilerId -> UnitId
+                       -> IO a -> IO a
+withIncomingUnitIdLock StoreDirLayout{storeIncomingLock} compid unitid action = do
+    bracket
+      (do h <- openFile lockFile ReadWriteMode; hLock h ExclusiveLock; return h)
```

Comment:
Occasionally on Windows, bad process control can result in zombie cabal.exe processes living way longer than you expect them to (this generally doesn't happen to me on Linux.) If this occurs, and the process had taken out a file lock, then all other cabal processes will block perpetually trying to take the lock. This sounds very confusing to debug.

The easiest way I can think to make it easier to debug in this situation is to add logging, emitting when we are taking out a lock, and on which file. This way if you rerun cabal in verbose mode you can see that it hangs immediately after trying to take the lock.